go-mode.el

```diff
@@ -1562,6 +1568,143 @@ for."
       (if (not (eq cur-buffer (current-buffer)))
           (display-buffer (current-buffer) `(,go-coverage-display-buffer-func))))))
 
+(define-prefix-command 'go-goto-map)
+(define-key go-mode-map (kbd "C-c g") 'go-goto-map)
+
+(defun go-goto-function ()
+  "Go to the function defintion above point.
+
+If we are on a docstring, follow the docstring down.
+If no function is found, assume that we are at the top of a file
+and search forward instead."
+  (interactive)
+
+  (beginning-of-line)
+  (cond
+   ((looking-at "^//")
+    ;; In case we are looking at the docstring, move on forward until we are not anymore
+    (while (looking-at "^//")
+      (forward-line 1)))
+   ((not (looking-at "^func"))
+    ;; If we are not looking at the beginning of a function line, do a regexp search backwards
+    (re-search-backward "\\<func\\>" nil t)
+
+    (when (not (looking-at "func"))
+      (re-search-forward "\\<func\\>" nil t)
+      (backward-word)))))
+
+(defun go-goto-function-name ()
+  "Go to the name of the current function.
+
+If the function is a test, place point after 'Test'.
+If the function is anonymous, place point on the 'func' keyword."
+  (interactive)
+  (when (not (looking-at "func"))
+    (go-goto-function))
+  ;; If we are looking at func( we are on an anonymous function and
+  ;; nothing else should be done.
+  (when (not (looking-at "func("))
+    (let ((words 1)
+          (chars 1))
+      (when (looking-at "func (")
+        (setq words 3
+              chars 2))
+      (forward-word words)
+      (forward-char chars)
+      (when (looking-at "Test")
+        (forward-char 4)))))
+
+(define-key go-goto-map (kbd "f") 'go-goto-function-name)
```

Comment:
Please move all `define-key` to a single place. Optimally below the definition of go-mode-map.